File: pythonlibrary/aibrd/parsers/confluence.py
"""Confluence page ingestion — fetch BRD content directly from Confluence Cloud or Server/DC."""

from __future__ import annotations
import base64
import re
import urllib.request
import urllib.parse
import json
from dataclasses import dataclass

from ..models.brd import RawBRD


@dataclass
class ConfluenceConfig:
    base_url: str        # e.g. https://yourorg.atlassian.net
    space_key: str       # e.g. ENG
    token: str           # API token or PAT
    page_id: str | None = None
    page_title: str | None = None
    email: str | None = None   # Atlassian Cloud: email + token; Server/DC: token only


def _auth_header(cfg: ConfluenceConfig) -> str:
    if cfg.email:
        creds = base64.b64encode(f"{cfg.email}:{cfg.token}".encode()).decode()
        return f"Basic {creds}"
    return f"Bearer {cfg.token}"


def _fetch_json(url: str, auth: str) -> dict:
    req = urllib.request.Request(url, headers={"Authorization": auth, "Accept": "application/json"})
    with urllib.request.urlopen(req, timeout=30) as resp:
        return json.loads(resp.read().decode())


def _strip_html(html: str) -> str:
    html = re.sub(r"<br\s*/?>", "\n", html, flags=re.IGNORECASE)
    html = re.sub(r"</p>", "\n\n", html, flags=re.IGNORECASE)
    html = re.sub(r"</li>", "\n", html, flags=re.IGNORECASE)
    html = re.sub(r"</h[1-6]>", "\n", html, flags=re.IGNORECASE)
    html = re.sub(r"<[^>]+>", "", html)
    html = html.replace("&amp;", "&").replace("&lt;", "<").replace("&gt;", ">").replace("&nbsp;", " ")
    return re.sub(r"\n{3,}", "\n\n", html).strip()


def _resolve_page_id(cfg: ConfluenceConfig, auth: str) -> str:
    if cfg.page_id:
        return cfg.page_id
    title = urllib.parse.quote(cfg.page_title or "")
    url = f"{cfg.base_url}/wiki/rest/api/content?spaceKey={cfg.space_key}&title={title}&expand=version"
    data = _fetch_json(url, auth)
    results = data.get("results", [])
    if not results:
        raise ValueError(f"No Confluence page found titled '{cfg.page_title}' in space '{cfg.space_key}'")
    return results[0]["id"]


def ingest_confluence_page(cfg: ConfluenceConfig) -> RawBRD:
    """Fetch a Confluence page (+ direct children) and return as RawBRD."""
    auth = _auth_header(cfg)
    page_id = _resolve_page_id(cfg, auth)

    url = f"{cfg.base_url}/wiki/rest/api/content/{page_id}?expand=body.storage,version,ancestors"
    page = _fetch_json(url, auth)
    title = page["title"]
    body = _strip_html(page["body"]["storage"]["value"])
    text = f"# {title}\n\n{body}"

    # Pull direct children
    children_url = f"{cfg.base_url}/wiki/rest/api/content/{page_id}/child/page?expand=body.storage&limit=50"
    try:
        children_data = _fetch_json(children_url, auth)
        for child in children_data.get("results", []):
            child_body = _strip_html(child["body"]["storage"]["value"])
            text += f"\n\n## {child['title']}\n\n{child_body}"
    except Exception:
        pass  # children optional

    source = f"confluence:{cfg.base_url}/wiki/spaces/{cfg.space_key}/pages/{page_id}"
    filename = re.sub(r"[^a-zA-Z0-9]", "_", title) + ".confluence"
    return RawBRD(text=text, source=source, filename=filename)

File: vscode-extension/src/core/parsers/confluence.ts
import * as https from 'https'
import * as http from 'http'
import { RawBRD } from '../models/brd'

export interface ConfluenceConfig {
  baseUrl: string        // e.g. https://yourorg.atlassian.net
  spaceKey: string       // e.g. ENG
  pageId?: string        // specific page ID, or use title search
  pageTitle?: string     // search by title if pageId not given
  token: string          // Confluence API token (PAT or Bearer)
  email?: string         // for cloud: Basic auth email (token = API key)
}

interface ConfluencePage {
  id: string
  title: string
  body: { storage: { value: string } }
  version: { number: number }
  ancestors: Array<{ id: string; title: string }>
}

function fetchJson(url: string, headers: Record<string, string>): Promise<unknown> {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url)
    const mod = parsed.protocol === 'https:' ? https : http
    const req = mod.request(
      { hostname: parsed.hostname, path: parsed.pathname + parsed.search, headers },
      res => {
        let data = ''
        res.on('data', chunk => (data += chunk))
        res.on('end', () => {
          try { resolve(JSON.parse(data)) }
          catch { reject(new Error(`Non-JSON response from Confluence: ${data.slice(0, 200)}`)) }
        })
      }
    )
    req.on('error', reject)
    req.end()
  })
}

function stripHtml(html: string): string {
  return html
    .replace(/<br\s*\/?>/gi, '\n')
    .replace(/<\/p>/gi, '\n\n')
    .replace(/<\/li>/gi, '\n')
    .replace(/<\/h[1-6]>/gi, '\n')
    .replace(/<[^>]+>/g, '')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&nbsp;/g, ' ')
    .replace(/\n{3,}/g, '\n\n')
    .trim()
}

async function resolvePageId(cfg: ConfluenceConfig, authHeader: string): Promise<string> {
  if (cfg.pageId) return cfg.pageId

  const title = encodeURIComponent(cfg.pageTitle ?? '')
  const url = `${cfg.baseUrl}/wiki/rest/api/content?spaceKey=${cfg.spaceKey}&title=${title}&expand=version`
  const res = await fetchJson(url, { Authorization: authHeader, Accept: 'application/json' }) as {
    results: Array<{ id: string }>
  }

  if (!res.results?.length) {
    throw new Error(`Confluence: no page found titled "${cfg.pageTitle}" in space "${cfg.spaceKey}"`)
  }
  return res.results[0].id
}

async function fetchChildPages(
  cfg: ConfluenceConfig,
  pageId: string,
  authHeader: string
): Promise<ConfluencePage[]> {
  const url = `${cfg.baseUrl}/wiki/rest/api/content/${pageId}/child/page?expand=body.storage,version,ancestors&limit=50`
  const res = await fetchJson(url, { Authorization: authHeader, Accept: 'application/json' }) as {
    results: ConfluencePage[]
  }
  return res.results ?? []
}

/**
 * Ingest a Confluence page (and its direct children) into a RawBRD.
 * Handles both Confluence Cloud (email+token → Basic) and Server/DC (token → Bearer).
 */
export async function ingestConfluencePage(cfg: ConfluenceConfig): Promise<RawBRD> {
  const authHeader = cfg.email
    ? `Basic ${Buffer.from(`${cfg.email}:${cfg.token}`).toString('base64')}`
    : `Bearer ${cfg.token}`

  const pageId = await resolvePageId(cfg, authHeader)
  const url = `${cfg.baseUrl}/wiki/rest/api/content/${pageId}?expand=body.storage,version,ancestors`
  const page = await fetchJson(url, { Authorization: authHeader, Accept: 'application/json' }) as ConfluencePage

  let text = `# ${page.title}\n\n${stripHtml(page.body.storage.value)}`

  // Also pull direct children to get sub-sections
  const children = await fetchChildPages(cfg, pageId, authHeader)
  for (const child of children) {
    text += `\n\n## ${child.title}\n\n${stripHtml(child.body.storage.value)}`
  }

  const source = `confluence:${cfg.baseUrl}/wiki/spaces/${cfg.spaceKey}/pages/${pageId}`
  return {
    text,
    source,
    filename: `${page.title.replace(/[^a-zA-Z0-9]/g, '_')}.confluence`,
  }
}
